systemtests/tests/test_direct_upload.py

```diff
@@ -0,0 +1,155 @@
+pytestmark = pytest.mark.upload
+
+small_archive_param = pytest.param((2**19, "windows"), id="small")
+large_archive_param = pytest.param(
+    (2**30, "linux"), id="large", marks=pytest.mark.large_files
+)
+
+small_and_large_archives = pytest.mark.parametrize(
+    "zip_archive", [small_archive_param, large_archive_param], indirect=True
+)
+small_archive = pytest.mark.parametrize(
+    "zip_archive", [small_archive_param], indirect=True
+)
+
+
+class TestDirectUpload:
+    # Tests within a class scope are executed in definition order.
+
+    @small_and_large_archives
+    def test_upload(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        response = tecken_client.upload(zip_archive.file_name)
+        assert response.status_code == 201
+        zip_archive.uploaded = True
+
+    @small_and_large_archives
+    def test_download(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        for sym_file in zip_archive.members:
+            response = tecken_client.download(sym_file.key())
+            assert response.status_code == 200
+            [redirect] = response.history
+            assert redirect.status_code == 302
+
+    @small_archive
+    def test_head_request(
+        self, tecken_client: TeckenClient, zip_archive: FakeZipArchive
+    ):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        for sym_file in zip_archive.members:
+            response = tecken_client.download(sym_file.key(), method="HEAD")
+            assert response.status_code == 200
+            assert not response.history
+
+    @small_archive
+    @pytest.mark.nginx
+    def test_headers(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        # These are exclusively security headers added by nginx
+        expected_tecken_headers = {
+            "X-Content-Type-Options": "nosniff",
+            "X-Frame-Options": "DENY",
+            "Content-Security-Policy": ANY,
+            "Strict-Transport-Security": "max-age=31536000",
+        }
+        # These are headers sent by the storage backend (S3 or GCS)
+        expected_storage_headers = {
+            "Content-Encoding": "gzip",
+            "Content-Length": ANY,
+        }
+        response = tecken_client.download(zip_archive.members[0].key())
+        [redirect] = response.history
+
+        actual_tecken_headers = {
+            name: redirect.headers[name] for name in expected_tecken_headers
+        }
+        assert actual_tecken_headers == expected_tecken_headers
+
+        actual_storage_headers = {
+            name: response.headers[name] for name in expected_storage_headers
+        }
+        assert actual_storage_headers == expected_storage_headers
+
+    @small_archive
+    def test_code_info_lookup(
```

Comment:
This test should only use Windows sym files. Linux sym files and macOS sym files don't have code file data.

Reply:
We talked about this. I want to adjust the structure here so it's easier to see that. We'll tinker with that in a followup PR.